src/katgpucbf/xbgpu/bsend.py

```diff
@@ -0,0 +1,429 @@
+        This should only be done when :attr:`future` is done.
+        """
+        return self._timestamp
+
+    @timestamp.setter
+    def timestamp(self, value: int) -> None:
+        delta = value - self.timestamp
+        self._timestamps += delta
+        self._timestamp = value
+
+    def send(
+        self,
+        send_stream: "BSend",
+        time_converter: TimeConverter,
+        sensors: SensorSet,
+    ) -> asyncio.Future:
+        """
+        Transmit a chunk's heaps over a SPEAD stream.
+
+        This method returns immediately and sends the data asynchronously. Before
+        modifying the chunk, first await :attr:`future`.
+
+        .. todo::
+
+            Also update its relevant counters and sensor values.
+        """
+        if any(send_stream.tx_enabled):
+            send_futures = []
+            for frame in self._frames:
+                heaps_to_send = [heap for heap, enabled in zip(frame.heaps, send_stream.tx_enabled) if enabled]
+                send_futures.append(
+                    send_stream.stream.async_send_heaps(heaps_to_send, mode=spead2.send.GroupMode.ROUND_ROBIN)
+                )
+            # TODO: Update counters and sensor with chunk.saturation
+            self.future = asyncio.gather(*send_futures)
+        else:
+            # TODO: Is it necessary to handle this case?
+            self.future = asyncio.create_task(send_stream.stream.async_flush())
+        return self.future
+
+
+class BSend:
+    """
+    Class for turning tied array channelised voltage products into SPEAD heaps.
+
+    This class creates a queue of chunks that can be sent out onto the network.
+    To obtain a chunk, call :meth:`get_free_chunk` - which will return a
+    :class:`Chunk`. This object will create a limited number of transmit
+    buffers and keep recycling them, avoiding any memory allocation at runtime.
+
+    The transmission of a chunk's data is abstracted by :meth:`send_chunk`. This
+    invokes transmission and immediately returns the :class:`Chunk` back to the
+    queue for reuse.
+
+    This object keeps track of each tied-array-channelised-voltage data stream by
+    means of a substreams in :class:`spead2.send.asyncio.AsyncStream`, allowing
+    for individual enabling and disabling of the data product.
+
+    To allow this class to be used with multiple transports, the constructor
+    takes a factory function to create the stream.
+
+    Parameters
+    ----------
+    outputs
+        Sequence of :class:`.output.BOutput`.
+    heaps_per_fengine_per_chunk
+        Number of SPEAD heaps from one F-engine in a single received Chunk.
+    n_tx_items
+        Number of :class:`Chunk` to create.
+    adc_sample_rate, n_channels, n_channels_per_substream, spectra_per_heap, channel_offset
+        See :class:`.XBEngine` for further information.
+    timestamp_step
+        The timestamp step between successive heaps, as dictated by the XBEngine.
```

Comment:
```suggestion
        The timestamp step between successive heaps.
```
It's dictated by the user of this class, which could be XBEngine, or it could be test_bsend.